File: longest_consequtive.py
from typing import List

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        # A set to represent the characters currently in our window (O(1) lookup)
        char_set = set()
        
        # 'left' is the starting index of the current window
        left = 0
        
        # 'max_length' tracks the best result found so far
        max_length = 0

        # 'right' is the index being added to the window (represented by 'right' pointer)
        for right in range(len(s)):
            
            # 1. Collision Check (Shrink the window if the new character is a duplicate)
            # If the character s[right] is already in the set, we must remove elements 
            # from the left side until the duplicate is gone.
            while s[right] in char_set:
                
                # Remove the leftmost character from the set
                char_set.remove(s[left])
                
                # Move the left pointer one step to the right
                left += 1
                
            # 2. Expansion (Add the new, non-repeating character and update max length)
            
            # Add the current right character to the set (it's guaranteed to be unique now)
            char_set.add(s[right])
            
            # Calculate the length of the current window: right - left + 1
            # And update the overall maximum length
            max_length = max(max_length, right - left + 1)
            
        return max_length

File: longest_consequtive.cpp
#include <string>
#include <unordered_map>
#include <algorithm> // For std::max
#include <vector>    // For std::vector (though not strictly needed here, often used with LeetCode)

// Using std namespace for convenience in this context
using namespace std;

class Solution {
public:
    int characterReplacement(string s, int k) {
        
        int left = 0;
        
        // Dictionary/Hash Map equivalent: stores character frequencies
        unordered_map<char, int> freq;
        int res = 0;

        // The 'right' pointer expands the sliding window
        for (int right = 0; right < s.length(); ++right) {
            
            // Python's try/except for freq[s[right]]+=1 is achieved idiomatically 
            // in C++ by map's auto-initialization. If the key doesn't exist, 
            // it's initialized to 0, then incremented to 1.
            freq[s[right]]++;

            int curlen = right - left + 1; // Finding total number of elements in the current sliding window

            // --- Translating max(freq.values()) ---
            // This loop ensures we find the maximum frequency (maxF) in the current window.
            // Note: This is an O(N) operation inside the main O(N) loop, making this solution O(N^2).
            int maxF = 0;
            for (auto const& [key, val] : freq) {
                maxF = max(maxF, val);
            }
            
            // Check if the number of characters needing replacement (curlen - maxF) 
            // exceeds the allowed limit (k).
            if ((curlen - maxF) > k) {
                
                // 1. Decrement the frequency of the character leaving the window (s[left]).
                // The C++ logic preserves the original Python intention: decrement BEFORE moving left.
                freq[s[left]]--;
                
                // 2. Slide the window (move the left pointer one step to the right).
                left++;
            }

            // Update the maximum length found so far.
            // Since the window only grows or stays the same size, we update here.
            res = max(res, right - left + 1);
        }

        return res;
    }
};
